File: closure.nim

proc current_timestamp*(): int64 {.importc, cdecl, dynlib:"libnewplus.so".}

type
  GetSet = tuple[
    get: proc(): int32 {.closure.},
    set: proc(v: int32) {.closure.}]

proc run(gs: GetSet, loops: int64) = 
  while gs.get() < loops: 
    let x:int32 = gs.get() + 1
    gs.set(x)

proc newGetSet(): GetSet {.inline.} =
  var mv:int32 = 0
  
  result = (
    get: proc(): int32 = 
      result = mv,
    set: proc(v: int32) = 
      mv = v
  )

when isMainModule:

  var start, endts: int64
  
  let gs = newGetSet()

  start = current_timestamp()

  run(gs, 2_000_000_000)

  endts = current_timestamp()

  echo(endts - start)

File: closure.rs
use std::libc::*;

#[link_args = "-lnewplus"]
extern
{
    fn current_timestamp() -> c_long;
}

struct GetSet<'self>
{
    get: &'self fn() -> c_int,
    set: &'self fn(c_int)
}

#[fixed_stack_segment]
fn run(gs: &GetSet, loops: c_int)
{
    while (gs.get)() < loops
    {
        let x = (gs.get)() + 1;
        (gs.set)(x);
    }
}

#[fixed_stack_segment]
fn main()
{
    let mut start: c_long;
    let mut endts: c_long;
    
    // closure start
    let mut mv:c_int = 0;
    let fn_get = || -> c_int { mv };
    let fn_set = |v: c_int| { mv = v; };
    
    let gs = GetSet{get: fn_get, set: fn_set};
    // closure end

    unsafe { start = current_timestamp(); }

    // run
    run(&gs, 2000000000);

    unsafe { endts = current_timestamp(); }

    println!("{}", endts - start);
}
